src/middlewared/middlewared/plugins/cluster_linux/ctdb_public.py

```diff
@@ -0,0 +1,238 @@
+        if not await self.middleware.call('service.started', 'glusterd'):
+            verrors.add(
+                f'{schema_name}.glusterd',
+                'The "glusterd" service is not started.',
+            )
+
+        try:
+            mounted = pathlib.Path(SHARED_VOL).is_mount()
+        except Exception:
+            mounted = False
+
+        if not mounted:
+            verrors.add(
+                f'{schema_name}.{SHARED_VOL}',
+                f'"{SHARED_VOL}" is not mounted'
+            )
+
+        # we have to make sure the IP that was given to us doesn't exist
+        # on the node since ctdb daemon automatically creates and manages
+        # this address
+        node_ips = [i['address'] for i in (await self.middleware.call('interface.ip_in_use', {'static': True}))]
+        if data['ip'] in node_ips:
+            verrors.add(
+                f'{schema_name}.{data["ip"]}',
+                f'{data["ip"]} already exist on this node. ',
+            )
+
+        verrors.check()
+
+        # get the current ips in the cluster
+        cur_ips = [i['id'] for i in (await self.middleware.call('ctdb.private.ips.query'))]
+        cur_ips.extend([i['id'] for i in (await self.middleware.call('ctdb.public.ips.query'))])
+
+        if not delete:
+            # validate the netmask
+            netmask_min = 0
+            v4max = 32
+            v6max = 128
+            bad_netmask = False
+            if ':' in data['ip'] and data['netmask'] > v6max:
+                bad_netmask = True
+            elif data['netmask'] > v4max:
+                bad_netmask = True
+            elif data['netmask'] < netmask_min:
+                bad_netmask = True
+
+            if bad_netmask:
+                verrors.add(
+                    f'{schema_name}.{data["netmask"]}',
+                    f'The netmask: "{data["netmask"]}" for "{data["ip"]}" is invalid.',
+                )
+
+            # validate the interface
+            node_ints = [i['id'] for i in (await self.middleware.call('interface.query'))]
+            if data['interface'] not in node_ints:
+                verrors.add(
+                    f'{schema_name}.{data["interface"]}',
+                    f'{data["interface"]} does not exist on this node. ',
+                )
+
+            # make sure public ip doesn't already exist in the cluster
+            if data['ip'] in cur_ips:
+                verrors.add(
+                    f'{schema_name}.{data["ip"]}',
+                    f'"{data["ip"]}" already in the cluster.'
+                )
+        else:
+            # make sure public ip exists before deleting from the cluster
+            if data['ip'] not in cur_ips:
+                verrors.add(
+                    f'{schema_name}.{data["ip"]}',
+                    f'"{data["ip"]}" does not exist in the cluster.'
+                )
+
+        verrors.check()
+
+    @private
+    def update_file(self, data, verrors, delete=False):
```

Comment:
Not sure if this is important, may want to consult with William, but we may need an update() method that wraps around this.

Reply:
`do_update` methods don't make sense in this context.

For example. `update()` in the private IP API would be change `#192.168.1.150` to `192.168.1.150` so it wouldn't be clean to do it that way since I'd have to by pass `query()` and read directly from the file.